### PR TITLE
Balder/put commit params in struct

### PR DESCRIPTION
**Pull request by baldersheim**
@toregge or @geirst PR

### CODE REVIEW

File: searchcore/src/tests/proton/attribute/attribute_manager/attribute_manager_test.cpp [hunk 1]
```diff
@@ -395,7 +395,7 @@ TEST_F("require that predicate attributes are flushed and loaded", BaseFixture)
         uint32_t doc_id;
         a1->addDoc(doc_id);
         index.indexEmptyDocument(doc_id);
-        pa.commit(100, 100);
+        pa.commit(CommitParam(100, true));
```
Comment (geirst) [inline searchcore/src/tests/proton/attribute/attribute_manager/attribute_manager_test.cpp:398]: Why are we using _forceUpdateStats=true_ here? That is not equal to the previous behaviour. The same comment applies for other changes in this file, attribute_test.cpp and attributeflush_test.cpp.

File: searchcore/src/tests/proton/attribute/attributeflush_test.cpp [hunk 2]
```diff
@@ -479,13 +479,13 @@ Test::requireThatOnlyOneFlusherCanRunAtTheSameTime()
...
     av->addDocs(10000);
-    av->commit(9,9);
+    av->commit(CommitParam(9, true));
     IFlushTarget::SP ft = am.getFlushable("a8");
     (static_cast<FlushableAttribute *>(ft.get()))->setCleanUpAfterFlush(false);
     vespalib::ThreadStackExecutor exec(16, 64000);
 
     for (size_t i = 10; i < 100; ++i) {
-        av->commit(i, i);
+        av->commit(CommitParam(i));
```
Comment (geirst) [inline searchcore/src/tests/proton/attribute/attributeflush_test.cpp:488]: This is the only place we use  _forceUpdateStats=false_, as before. Why is that?

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_initializer.cpp [hunk 3]
```diff
@@ -198,7 +199,7 @@ AttributeInitializer::loadAttribute(const AttributeVectorSP &attr,
             attr->getBaseFileName().c_str());
         return false;
     } else {
-        attr->commit(serialNum, serialNum);
+        attr->commit(CommitParam(serialNum, true));
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/attribute/attribute_initializer.cpp:202]: Why are we using _forceUpdateStats=true_ here? That is not equal to the previous behaviour. The same comment applies for other changes in this file.

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.cpp [hunk 4]
```diff
@@ -208,11 +209,11 @@ applyCompactLidSpace(uint32_t wantedLidLimit, SerialNum serialNum, AttributeVect
          * later config changes removing the attribute then it might
          * be smaller than expected during transaction log replay.
          */
-        attr.commit();
+        attr.commit(true);
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.cpp:212]: Why are we using _forceUpdateStats=true_ here? That is not equal to the previous behaviour. The same comment applies for the change further down.

File: searchcore/src/vespa/searchcore/proton/attribute/flushableattribute.cpp [hunk 5]
```diff
@@ -62,7 +62,7 @@ FlushableAttribute::Flusher::Flusher(FlushableAttribute & fattr, SerialNum syncT
       _syncToken(syncToken),
       _flushFile("")
 {
-    fattr._attr->commit(syncToken, syncToken);
+    fattr._attr->commit(CommitParam(syncToken, true));
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/attribute/flushableattribute.cpp:65]: Why are we using _forceUpdateStats=true_ here? That is not equal to the previous behaviour.

File: searchcore/src/vespa/searchcore/proton/documentmetastore/documentmetastoreinitializer.cpp [hunk 6]
```diff
@@ -49,7 +49,7 @@ DocumentMetaStoreInitializer::run()
             if (!_dms->load()) {
                 throw IllegalStateException(failedMsg(_docTypeName.c_str()));
             } else {
-                _dms->commit(snap.syncToken, snap.syncToken);
+                _dms->commit(search::CommitParam(snap.syncToken, true));
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/documentmetastore/documentmetastoreinitializer.cpp:52]: Why are we using _forceUpdateStats=true_ here? That is not equal to the previous behaviour.

File: searchcore/src/vespa/searchcore/proton/server/documentdb.cpp [hunk 7]
```diff
@@ -338,7 +338,7 @@ DocumentDB::enterOnlineState()
 {
     // Called by executor thread
     // Ensure that all replayed operations are committed to memory structures
-    _feedView.get()->forceCommit(_feedHandler->getSerialNum());
+    _feedView.get()->forceCommit(CommitParam(_feedHandler->getSerialNum(), true));
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/documentdb.cpp:341]: Why are we using _forceUpdateStats=true_ here? That is not equal to the previous behaviour. The same comment applies for other changes in this file.

File: searchcore/src/vespa/searchcore/proton/server/feedhandler.cpp [hunk 8]
```diff
@@ -547,7 +548,7 @@ FeedHandler::initiateCommit() {
     if (_activeFeedView) {
         using KeepAlivePair = vespalib::KeepAlive<std::pair<CommitResult, DoneCallback>>;
         auto pair = std::make_pair(std::move(commitResult), std::move(onCommitDoneContext));
-        _activeFeedView->forceCommit(_serialNum, std::make_shared<KeepAlivePair>(std::move(pair)));
+        _activeFeedView->forceCommit(CommitParam(_serialNum, false), std::make_shared<KeepAlivePair>(std::move(pair)));
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/feedhandler.cpp:551]: No need to use explicit false here. Could use _CommitParam(_serialNum)_.

File: searchcore/src/vespa/searchcore/proton/server/lid_space_compaction_handler.cpp [hunk 9]
```diff
@@ -93,7 +94,7 @@ LidSpaceCompactionHandler::handleCompactLidSpace(const CompactLidSpaceOperation
 {
     assert(_subDb.sub_db_id() == op.getSubDbId());
     _subDb.feed_view()->handleCompactLidSpace(op);
-    _subDb.feed_view()->forceCommit(op.getSerialNum(), std::move(compact_done_context));
+    _subDb.feed_view()->forceCommit(CommitParam(op.getSerialNum(), true), std::move(compact_done_context));
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/lid_space_compaction_handler.cpp:97]: Why are we using _forceUpdateStats=true_ here? That is not equal to the previous behaviour.

File: searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp [hunk 10]
```diff
@@ -819,10 +817,12 @@ StoreOnlyFeedView::handleCompactLidSpace(const CompactLidSpaceOperation &op)
                                                                 _gidToLidChangeHandler.grab_pending_changes(),
                                                                 DoneCallback()));
         commitContext->holdUnblockShrinkLidSpace();
-        internalForceCommit(serialNum, commitContext);
+        internalForceCommit(CommitParam(serialNum, true), commitContext);
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp:820]: Why are we using _forceUpdateStats=true_ here? That is not equal to the previous behaviour.

File: searchcore/src/vespa/searchcore/proton/test/attribute_utils.h [hunk 11]
```diff
@@ -18,7 +18,7 @@ struct AttributeUtils
         for (uint32_t i = 1; i < ia.getNumDocs(); ++i) {
             ia.update(i, value);
         }
-        ia.commit(lastSyncToken, lastSyncToken);
+        ia.commit(search::CommitParam(lastSyncToken, true));
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/test/attribute_utils.h:21]: Why are we using _forceUpdateStats=true_ here? That is not equal to the previous behaviour. The same comment applies for other changes in this file.

File: searchlib/src/vespa/searchlib/common/commit_param.h [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+{
+    CommitParam(SerialNum serialNum) noexcept : CommitParam(serialNum, false) {}
+    CommitParam(SerialNum serialNum, bool forceUpdateStats) noexcept : CommitParam(serialNum, serialNum, forceUpdateStats) {}
+    CommitParam(SerialNum firstSerialNum, SerialNum lastSerialNum, bool forceUpdateStats) noexcept
+        : _firstSerialNum(firstSerialNum),
+          _lastSerialNum(lastSerialNum),
+          _forceUpdateStats(forceUpdateStats)
+    {}
+
+    const SerialNum _firstSerialNum;
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/common/commit_param.h:21]: Consider hiding member variables, and use a class instead.

**Review by geirst — COMMENTED** [review]
Overall looks good, but have a lot of questions regarding why _forceUpdateStats_ is changed to _true_ in various places.

File: searchlib/src/vespa/searchlib/common/commit_param.h [hunk 13]
```diff
@@ -0,0 +1,26 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/common/commit_param.h:1]: Done

File: searchlib/src/vespa/searchlib/common/commit_param.h (hunk 12 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/common/commit_param.h:21]: Fixed

File: searchcore/src/tests/proton/attribute/attribute_manager/attribute_manager_test.cpp (hunk 1 above)
Comment (baldersheim) [reply searchcore/src/tests/proton/attribute/attribute_manager/attribute_manager_test.cpp:398]: Removed semantic change

File: searchcore/src/tests/proton/attribute/attributeflush_test.cpp (hunk 2 above)
Comment (baldersheim) [reply searchcore/src/tests/proton/attribute/attributeflush_test.cpp:488]: Removed semantic change

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_initializer.cpp (hunk 3 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/attribute/attribute_initializer.cpp:202]: Removed semantic change

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.cpp (hunk 4 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.cpp:212]: Removed semantic change

File: searchcore/src/vespa/searchcore/proton/server/feedhandler.cpp (hunk 8 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/feedhandler.cpp:551]: As false is a performance optimization, it is better to make the optimization explicit and default to non-optimized.

File: searchcore/src/vespa/searchcore/proton/server/lid_space_compaction_handler.cpp (hunk 9 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/lid_space_compaction_handler.cpp:97]: Removed semantic change

File: searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp (hunk 10 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp:820]: Removed semantic change

File: searchcore/src/vespa/searchcore/proton/test/attribute_utils.h (hunk 11 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/test/attribute_utils.h:21]: Removed semantic change

**Comment by baldersheim** [comment]
The changes are are done in #16705 which superceeds this one.

**Pull request closed without merging**